### PR TITLE
Add infection for mutation testing, various cleanups of the test suite and improve test coverage

### PR DESCRIPTION
**Pull request by ben-challis**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,15 +16,15 @@
   },
   "autoload-dev": {
     "psr-4": {
-      "Tests\\Lendable\\Clock\\Unit\\": "tests/unit/",
-      "Tests\\Lendable\\Clock\\Functional\\": "tests/functional/"
+      "Tests\\Lendable\\Clock\\Unit\\": "tests/unit/"
```
Comment (ben-challis) [inline composer.json:19]: From template, there's no functional tests as a vfs is used for testing `PersistedFixedClock` as a unit test instead.

File: infection.json.dist [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    "source": {
+        "directories": [
+            "lib"
+        ]
+    },
+    "mutators": {
+        "@default": true,
+        "IncrementInteger": {
+            "ignoreSourceCodeByRegex": [
+                ".*json_decode.*"
```
Comment (ben-challis) [inline infection.json.dist:12]: `512` is specified as we need to specify `\JSON_THROW_ON_ERROR`, Infection wants to mutate this but it's difficult to test the impact this has due to the data we're serializing.

File: lib/Date.php [hunk 3]
```diff
@@ -47,10 +47,12 @@ public static function fromYearMonthDayString(string $value): self
             throw new InvalidDate('Failed to parse string as a Y-m-d formatted date.');
         }
 
+        [, $year, $month, $day] = $matches;
```
Comment (ben-challis) [inline lib/Date.php:50]: Infection changes `$matches[1]` to `$matches[0]`, which is the entire match string, and `2020-01-02` casts to `2020` as an int so is uncaught but also impossible to detect, unless I'm missing something. Seems overkill to be validating each value in the matches is an integer due to the regex pattern already doing this.

File: tests/unit/DateTest.php [hunk 4]
```diff
@@ -130,15 +131,30 @@ public function it_can_be_constructed_from_a_formatted_string(
         $this->assertSame($expectedStringRepresentation, $result->toYearMonthDayString());
     }
 
+    /**
+     * @return iterable<array{string}>
+     */
+    public function provideInvalidYearMonthDayStrings(): iterable
```
Comment (ben-challis) [inline tests/unit/DateTest.php:137]: Infection removes `$` from the regex and that was uncovered previously.

File: tests/unit/DateTest.php [hunk 5]
```diff
@@ -419,4 +435,15 @@ public function it_can_be_offset_by_days(int $days): void
 
         $this->assertTrue($expectedDate->equals($offsetDate));
     }
+
+    /**
+     * @test
+     */
+    public function it_can_be_constructed_from_a_number_of_years_ago(): void
```
Comment (ben-challis) [inline tests/unit/DateTest.php:442]: Need to consider what we do with this method as it's using the current timestamp which doesn't align well with the rest of the library, consider changing this to `Date::fromYearsAgo(Clock $c, int $years): Date` potentially? Or drop it?

File: tests/unit/PersistedFixedClockTest.php [hunk 6]
```diff
@@ -80,14 +80,22 @@ public function it_can_change_the_time_to_a_new_fixed_value(): void
...
         $updatedTime = $now->modify('+30 minutes');
 
         $clock->changeTimeTo($updatedTime);
+        $expectedTimestamp = '2021-05-05T14:41:49.128311';
+
+        $this->assertSame($expectedTimestamp, $clock->now()->format($timeFormat));
+        $this->assertSame($expectedTimestamp, $clock->nowMutable()->format($timeFormat));
+
+        // Verify that this value was persisted.
+        $clock = PersistedFixedClock::fromPersisted($vfs->url(), $fileNameGenerator);
```
Comment (ben-challis) [inline tests/unit/PersistedFixedClockTest.php:95]: Previously uncovered and only in-memory state checked.

File: tests/unit/Serialization/FastestTestChannelFileNameGeneratorTest.php [hunk 7]
```diff
@@ -14,12 +14,33 @@ final class FastestTestChannelFileNameGeneratorTest extends TestCase
...
+        $this->exerciseAndResetEnv(function (): void {
+            \putenv(\sprintf('%s=3', EnvCommandCreator::ENV_TEST_CHANNEL_READABLE));
+            $this->assertSame('now_3.json', (new FastestTestChannelFileNameGenerator())->generate());
+        });
+    }
+
+    /**
+     * @test
+     */
+    public function it_defaults_to_1_when_the_env_var_is_not_provided(): void
```
Comment (ben-challis) [inline tests/unit/Serialization/FastestTestChannelFileNameGeneratorTest.php:27]: Didn't cover lack of env var being set and falling back to `"1"`.

File: composer.json [hunk 8]
```diff
@@ -78,14 +81,15 @@
     "phpunit:unit": [
       "phpunit --colors --testsuite=unit --do-not-cache-result"
     ],
-    "phpunit:integration": [
```
Comment (ben-challis) [inline composer.json:81]: Didn't exist.

File: composer.json [hunk 9]
```diff
@@ -78,14 +81,15 @@
...
-      "phpunit --colors --testsuite=integration --do-not-cache-result"
+    "infection": [
+      "./bin/infection --threads=8 --only-covering-test-cases --min-msi=99 --show-mutations"
```
Comment (ben-challis) [inline composer.json:85]: Hard coding to 8 here as various Alpine images / OS setups may not have `nproc` installed and the test suite is fast.

Comment (marmichalski) [reply composer.json:85]: But GH worker has only 2 cores :panda_face:

Comment (ben-challis) [reply composer.json:85]: Doesn't seem to make it much slower - and much faster locally

File: tests/unit/DateTest.php (hunk 5 above)
Comment (marmichalski) [reply tests/unit/DateTest.php:442]: Good for now; drop the method later; implement `offsetByYears` (similar to already existing `offsetByDays`); be happy :crab:

**Review by marmichalski — APPROVED** [review]

**Pull request merged**